### PR TITLE
Further improve readAll(string) performance

### PR DESCRIPTION
**Pull request by mppf**
Follow-up to PR #24783.

This PR is intended to improve the performance of `readAll(string)` and other functions using `readStringImpl`. It adds `qio_channel_read_chars` which is similar to `qio_channel_read_char`, but can read many codepoints at once, and then uses it in `readStringImpl`.

How does it affect performance? These measurements were collected on my PC with `start_test -performance test/studies/shootout/submitted/regexredux2.chpl --numtrials 4` (and note that `main` here includes #24783). Please note that the performance gain for `readAll` itself will be a larger percentage (since it is not most of the benchmark time). The speed gain for `readAll(string)` specifically within the benchmark is about 2x.

| Benchmark    | main   | this PR | speedup    |
| ------------ | ------ | ------- | ---------- |
| regexredux2  | 1.42 s | 1.37 s  |  5% faster |

Reviewed by @jeremiah-corrado - thanks!

- [x] full comm=none testing

### CODE REVIEW

**Comment by mppf** [comment]
I've got these failures to look at

```
[Error matching compiler output for chpldoc/compflags/combinations/version.doc]
[Error matching compiler output for chpldoc/compflags/combinations/z_copy_version.doc]
[Error matching compiler output for chpldoc/compflags/combinations/z_copyright_help_version.doc]
[Error matching compiler output for chpldoc/compflags/combinations/z_help_version.doc]
[Error matching program output for library/standard/IO/advanceThrough/multiByteNotPresent]
[Error matching program output for library/standard/IO/advanceTo/multiByteNotPresent]
[Error matching program output for regex/ferguson/rechan]
[Error matching program output for regex/empty-channel-matches]
[Error matching program output for release/examples/benchmarks/shootout/revcomp-fast]
[Error matching program output for studies/shootout/reverse-complement/bharshbarg/revcomp-mark-skip-read-begin]
```

File: runtime/src/qio/qio_formatted.c [hunk 1]
```diff
@@ -4453,6 +4453,133 @@ qioerr _qio_channel_read_char_slow_unlocked(qio_channel_t* restrict ch, int32_t*
...
+      }
+    }
+
+    if (nCodepoints >= maxCodepoints) {
+      // stop if we have read the requested number of codepoints
+      break;
+    }
+
+    // read the rest of the codepoint if we are partway through one
+    while (state > 1 && nBytes < maxBytes) {
```
Comment (jeremiah-corrado) [inline runtime/src/qio/qio_formatted.c:4494]: I'm not seeing how it's possible to enter this loop. It looks like `state` will always have a value of `0` by this point. Am I missing something?

Comment (mppf) [reply runtime/src/qio/qio_formatted.c:4494]: (the buffer could have all the data in the file but often won't)

Comment (jeremiah-corrado) [reply runtime/src/qio/qio_formatted.c:4494]: Makes sense, thanks!

**Review by jeremiah-corrado — APPROVED** [review]

**Pull request merged**